app/coffee/Router.coffee

```diff
@@ -79,14 +79,30 @@ module.exports = Router =
 					if err?
 						Router._handleError null, err, client, "leaveProject"
 
-
-			client.on "joinDoc", (doc_id, fromVersion, callback) ->
-				# fromVersion is optional
-				if typeof fromVersion == "function"
+			# Variadic. The possible arguments:
+			# doc_id, callback
+			# doc_id, fromVersion, callback
+			# doc_id, options, callback
+			# doc_id, fromVersion, options, callback
+			client.on "joinDoc", (doc_id, fromVersion, options, callback) ->
+				if typeof fromVersion == "function" and !options
 					callback = fromVersion
 					fromVersion = -1
+					options = {}
+				else if typeof fromVersion == "number" and typeof options == "function"
+					callback = options
+					options = {}
+				else if typeof fromVersion == "object" and typeof options == "function"
+					callback = options
+					options = fromVersion
+					fromVersion = -1
+				else if typeof fromVersion == "number" and typeof options == "object"
+					# Called with 4 args, things are as expected
+				else
+					logger.error { arguments: arguments }, "unexpected arguments"
+					return # ????
```

Comment:
`callback` may not be defined here so we can rely on it to communicate the error back. So I wasn't sure what to do. WDYT?

Reply:
Hmm, good point. Do a best effort: `callback?(new Error(...))` (note the ? guard - my favourite CS feature that ES6 doesn't have!)

Reply:
Perhaps not for long https://github.com/tc39/proposal-optional-chaining

I'll fix that

Reply:
😀  That would make me very happy!